woocommerce/woocommerce-ios
# IPP: support card present refunds for Interac

Part of #5983 


### Description


For our IPP launch in Canada, we want to support Interac as one of the biggest payment providers in Canada. Normally, a refund can be issued via the API but Interac requires refunds to be made in person. This PR implemented the main flow for the refund for Interac, with a few follow-up tasks (subtasks of #5983) since this PR is already larger than I hoped (apology for the size! 🙇🏻‍♀️ ):
- Add unit tests for refund use case and orchestrator
- Look into why there is no "Remove card" and "Processing" messages from the reader service
- Update copy from payments to refunds
- Handle refund payment cancellation: `CardPresentModalReaderIsReady.didTapSecondaryButton` currently calls `CardPresentPaymentAction.cancelPayment` by default
- Retrying a failed in-person refund currently does not work, somehow `StripeCardReaderService.cancelRefund` never invokes the completion block 🤔 

### Major changes

- Created `CardPresentRefundOrchestrator` that orchestrates the lower-level actions needed for refunding in person
- Created `RefundSubmissionUseCase` (with a protocol `RefundSubmissionProtocol`) to encapsulate all the logic for submitting a refund whether it is in-person or not. It first checks whether in-person refund is required (Interac only so far), and initiates card reader connection and the following card-present refund after a reader is connected using `CardPresentRefundOrchestrator`. Both cases (in-person or not) submit the refund to the site in the end, with the difference that in-person flow does not show in-progress UI because card reader alerts UI is shown
- Updated `RefundConfirmationViewModel` and `RefundConfirmationViewController` accordingly

### Testing instructions


Please note the following:
- There are still a few pending tasks as mentioned above. If you encounter an issue, please check the pending tasks and make a comment if you don't see it logged!
- Since the API support is in-progress, the refund is currently submitted as "manual" (not to the original payment gateway) as a workaround

Prerequisites: the store is in Canada and eligible for IPP (setup instructions PdfdoF-D-p2)

#### In-person refund

- Go to the orders tab
- Create an order and pay it with Interac using a physical card reader, if you don't have one already
- Go to the order above
- Tap "Issue Refund" and select the refund amount to confirm --> the card reader connection alerts should appear
- Connect to WisePad 3 reader if you haven't
- Insert the Interac test card and enter pin number --> a success notice is shown with the updated order, and the refund should go through in wp-admin

#### Online refund

- Switch to another non-Canada store
- Go to the orders tab
- Create an order that is refundable, if you don't have one alrady
- Go to the order above
- Tap "Issue Refund" and select the refund amount to confirm --> a fullscreen in-progress UI should appear, and a success notice is shown with the updated order. The refund should also go through in wp-admin

### Screencast



https://user-images.githubusercontent.com/1945542/162394282-a9c4505d-364f-45c8-a770-da9aaf18daee.MP4



---
